modules/helfi_etusivu_entities/src/Plugin/Block/AnnouncementsBlock.php

```diff
@@ -0,0 +1,265 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Drupal\helfi_etusivu_entities\Plugin\Block;
+
+use Drupal\Component\Utility\Xss;
+use Drupal\Core\Block\Attribute\Block;
+use Drupal\Core\Cache\Cache;
+use Drupal\Core\Entity\ContentEntityInterface;
+use Drupal\Core\Entity\EntityInterface;
+use Drupal\Core\Entity\FieldableEntityInterface;
+use Drupal\Core\StringTranslation\TranslatableMarkup;
+use Drupal\helfi_etusivu_entities\Plugin\ExternalEntities\StorageClient\Announcements;
+use Drupal\node\Entity\Node;
+use Drupal\node\NodeInterface;
+
+/**
+ * Provides an 'Announcements' block.
+ */
+#[Block(
+  id: "announcements",
+  admin_label: new TranslatableMarkup("Announcements"),
+)]
+final class AnnouncementsBlock extends EtusivuEntityBlockBase {
+
+  public const VISIBILITY_ALL_WEIGHT = 0;
+  public const VISIBILITY_REGION_WEIGHT = 1;
+  public const VISIBILITY_PAGE_WEIGHT = 2;
+
+  /**
+   * Entity types supported by announcements and the corresponding field names.
+   */
+  private const ENTITY_TYPE_FIELDS = [
+    'node' => 'field_announcement_content_pages',
+    'tpr_unit' => 'field_announcement_unit_pages',
+    'tpr_service' => 'field_announcement_service_pages',
+  ];
+
+  /**
+   * {@inheritdoc}
+   */
+  protected function getLocalEntities(): array {
+    $langcodes = $this->getContentLangcodes();
+    $storage = $this->entityTypeManager->getStorage('node');
+
+    // Get all published announcement nodes.
+    $query = $storage->getQuery()
+      ->accessCheck(TRUE)
+      ->condition('type', 'announcement')
+      ->condition('status', NodeInterface::PUBLISHED)
+      ->condition('langcode', $langcodes, 'IN');
+
+    $fields = $this->entityFieldManager->getFieldDefinitions('node', 'survey');
+
+    // Query only local nodes.
+    if (isset($fields['field_publish_externally'])) {
+      $query->condition('field_publish_externally', FALSE);
+    }
+
+    return $storage->loadMultiple($query->execute());
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  protected function getRemoteEntities(): array {
+    $entityStorage = $this->getExternalEntityStorage('helfi_announcements');
+    $nodes = [];
+
+    /** @var \Drupal\external_entities\ExternalEntityInterface $announcement */
+    foreach ($entityStorage->loadMultiple() as $announcement) {
```

Comment:
`loadMultiple` should have better caching than the block that is cached per url per url & langcode.